### PR TITLE
[IMP] website, web_editor: add new chart snippet

### PR DESCRIPTION
**Pull request by fja-odoo**
Introducing a new snippet called Chart.
This snippet uses the lib https://www.chartjs.org/

6 types of graph are possible:
Bar vertical
Bar horizontal
Line
Pie
Doughnut
Radar

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
         }
     },
 });
+
+options.registry.InnerChart = options.Class.extend({
+    custom_events: _.extend({}, options.Class.prototype.custom_events, {
+        'get_custom_colors': '_onGetCustomColors',
+
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: What's that extra line?

File: addons/website/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
...
+    }),
+    events: _.extend({}, options.Class.prototype.events, {
+        'click we-button.add_column': '_onAddColumnClick',
+        'click we-button.add_row': '_onAddRowClick',
+        'click we-button.o_we_matrix_remove_col': '_onRemoveColumnClick',
+        'click we-button.o_we_matrix_remove_row': '_onRemoveRowClick',
+        'blur we-matrix input': '_onFocusOutMatrixInput',
+        'focus we-matrix input': '_onFocusMatrixInput',
+
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: ???

File: addons/website/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
...
+        'get_custom_colors': '_onGetCustomColors',
+
+    }),
+    events: _.extend({}, options.Class.prototype.events, {
+        'click we-button.add_column': '_onAddColumnClick',
+        'click we-button.add_row': '_onAddRowClick',
+        'click we-button.o_we_matrix_remove_col': '_onRemoveColumnClick',
+        'click we-button.o_we_matrix_remove_row': '_onRemoveRowClick',
+        'blur we-matrix input': '_onFocusOutMatrixInput',
+        'focus we-matrix input': '_onFocusMatrixInput',
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: I don't mind people other convention than me for this, but not in the same widget... above you write `_onXXXClick` and here `_onFocusXXX`. Put the event name at the end. Especially when reading "_onFocusMatrixInput", I imagine an input event into a "focus matrix" component

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
...
+
+    }),
+    themeArray: ['alpha', 'beta', 'gamma', 'delta', 'epsilon'],
+    /**
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Add an extra line feed between properties and the first method

File: addons/website/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
...
+    events: _.extend({}, options.Class.prototype.events, {
+        'click we-button.add_column': '_onAddColumnClick',
+        'click we-button.add_row': '_onAddRowClick',
+        'click we-button.o_we_matrix_remove_col': '_onRemoveColumnClick',
+        'click we-button.o_we_matrix_remove_row': '_onRemoveRowClick',
+        'blur we-matrix input': '_onFocusOutMatrixInput',
+        'focus we-matrix input': '_onFocusMatrixInput',
+
+    }),
+    themeArray: ['alpha', 'beta', 'gamma', 'delta', 'epsilon'],
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Maybe just initialize this in the init.

File: addons/website/static/src/js/widgets/website_chart.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+     * @override
+     */
+    start: function () {
+        const target = this.el.querySelector('canvas');
+        this._super.apply(this, arguments).then(() => {
+            const chart = new WebsiteChart(this);
+            chart.attachTo($(target));
+        });
+    }
+});
```
Comment (qsm-odoo) [inline addons/website/static/src/js/widgets/website_chart.js]: Why ? Why not directly adding the correct selector on the `WebsiteChart` widget ?

File: addons/website/static/src/js/widgets/website_chart.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    },
+    /**
+     *  Discard all library changes to reset the state of the Html.
+     */
+    destroy: function () {
+        const canvasEl = document.createElement('canvas');
+        for (const [key, value] of Object.entries(this.el.dataset)) {
+            canvasEl.dataset[key] = value;
+        }
+        this.$el.parent().empty().append($(canvasEl));
```
Comment (qsm-odoo) [inline addons/website/static/src/js/widgets/website_chart.js]: The library does not have a destroy method ? :o

File: addons/website/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -1940,4 +1941,447 @@ options.registry.SnippetMove = options.Class.extend({
...
+    /**
+     * @override
+     */
+    init: function () {
+        this._super.apply(this, arguments);
+        this.$widgetTarget = this.$target.closest('.s_chart');
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Reading your whole PR, I am not sure why you do not simply write the dataset on the `s_chart` element, do not retarget etc, and simply let the widget create the canvas the library needs and use the `s_char` dataset for the options... or is the library using the dataset on the canvas by itself?

File: addons/website/static/src/js/widgets/website_chart.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    },
+    /**
+     * @override
+     */
+    start: function () {
+        for (const [key, value] of Object.entries(this.options)) {
+            if (!this.el.dataset[key]) {
+                this.el.dataset[key] = typeof value === 'object' ? JSON.stringify(value) : value;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/widgets/website_chart.js]: Forgot to remove that... I used it at the beginning when the canvas could be blank at the instantiation.
Same for the options above it was there to instantiate a canvas from nothing which is not the case anymore.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -1940,4 +1941,436 @@ options.registry.SnippetMove = options.Class.extend({
...
+        setTimeout(() => this.lastEditableSelectedInput.focus());
+    },
+    /**
+     * @override
+     */
+    selectDataAttribute: function (previewMode, value, params) {
+        const dataName = params.attributeName;
+        const oldValue = this.$target[0].dataset[dataName];
+        this._super(...arguments);
+        if (this.$target[0].dataset[dataName] !== oldValue) {
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: You are right it is not really usefull here as the widget will be reloaded anyway.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -1940,4 +1941,436 @@ options.registry.SnippetMove = options.Class.extend({
...
+            if (dataName === 'type') {
+                this._setDefaultSelectedInput();
+                this.$target[0].dataset.data = this._matrixToChartData();
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: selectDataAttribute already reloads my graph so i only need to set the correct data on it and not reload it twice.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 12]
```diff
@@ -1940,4 +1941,436 @@ options.registry.SnippetMove = options.Class.extend({
...
+     * @override
+     */
+    updateUI: function () {
+        // Selected input might not be in dom anymore if col/row removed
+        if (!this.lastEditableSelectedInput.closest('table') || this.colorPaletteSelectedInput && !this.colorPaletteSelectedInput.closest('table')) {
+            this._setDefaultSelectedInput();
+        }
+        this._super(...arguments);
+        // prevent the columns from becoming too small.
+        this.tableEl.classList.toggle('o_we_matrix_five_col', !!this.tableEl.querySelector('tr td:nth-of-type(4)'));
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: Kind of the same but count might be simpler to understand

File: addons/website/static/src/js/editor/snippets.options.js [hunk 13]
```diff
@@ -1940,4 +1941,436 @@ options.registry.SnippetMove = options.Class.extend({
...
+        return newEl;
+    },
+    /**
+     * Display the remove button coresponding to the colIndex
+     * @private
+     * @param {Int} colIndex Can be undefined, if so the last remove button of the column will be shown
+     */
+    _displayRemoveColButton: function (colIndex) {
+        //Nbr of column minus header and button
+        const colCount = this.tableEl.querySelectorAll('tr:first-child th').length - 2;
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: .rows exists as it returns the TRs but no methods for the columns i think.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 14]
```diff
@@ -1940,4 +1941,436 @@ options.registry.SnippetMove = options.Class.extend({
...
+        }
+    },
+    /**
+     * @override
+     */
+    updateUI: function () {
+        // Selected input might not be in dom anymore if col/row removed
+        if (!this.lastEditableSelectedInput.closest('table') || this.colorPaletteSelectedInput && !this.colorPaletteSelectedInput.closest('table')) {
+            this._setDefaultSelectedInput();
+        }
```
[the remark this answers is not in the document]
Comment (fja-odoo) [reply addons/website/static/src/js/editor/snippets.options.js]: I need to set the right selected input before updating the UI. Else in _computeWidgetState the color picker will try to take the color from a non existing input.

File: addons/website/static/src/js/content/snippets.animation.js [hunk 15]
```diff
@@ -1148,6 +1148,128 @@ registry.anchorSlide = publicWidget.Widget.extend({
...
+        this._super.apply(this, arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * @private
+     * @param {*} color
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: `@param` with correct type
`@returns`

File: addons/website/static/src/js/content/snippets.animation.js [hunk 16]
```diff
@@ -1148,6 +1148,128 @@ registry.anchorSlide = publicWidget.Widget.extend({
...
+                    },
+                }],
+            };
+        }
+        const canvas = this.el.querySelector('canvas');
+        this.chart = new window.Chart(canvas, chartData);
+        return this._super.apply(this, arguments);
+    },
+    /**
+     *  Discard all library changes to reset the state of the Html.
```
Comment (qsm-odoo) [inline addons/website/static/src/js/content/snippets.animation.js]: `@override`

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 17]
```diff
@@ -408,6 +408,48 @@ body.editor_enable.editor_has_snippets {
...
+                        }
+                    }
+                    input {
+                        border: 1px solid var(--o-we-bg-color-dark);
+                        background-color: var(--o-we-bg-color-dark);
+                        color: var(--o-we-color);
+                        font-size: 12px;
+                        width: 100%;
+                    }
+                    &:last-child{
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss]: !

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 18]
```diff
@@ -408,6 +408,48 @@ body.editor_enable.editor_has_snippets {
             }
         }
 
+
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss]: !

File: addons/website/static/src/js/editor/snippets.options.js [hunk 19]
```diff
@@ -1940,4 +1941,450 @@ options.registry.SnippetMove = options.Class.extend({
...
+                    data.datasets[i].borderColor.push(el.dataset.borderColor || '');
+                }
+            });
+        });
+        return JSON.stringify(data);
+    },
+    /**
+     * Return a td containing a we-button with minus icon
+     *
+     * @param  {...string} classes Classes to add to the we-button
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: missing `@returns {HTMLElement}`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 20]
```diff
@@ -1940,4 +1941,450 @@ options.registry.SnippetMove = options.Class.extend({
...
+        trEl.appendChild(this._makeDeleteButton('o_we_matrix_remove_row'));
+        const tbody = this.tableEl.querySelector('tbody');
+        tbody.insertBefore(trEl, tbody.lastElementChild);
+    },
+    /**
+     * @private
+     * @param {string} tag tag of the HTML Element (td/th)
+     * @param {string} value The current value of the cell input
+     * @param {string} backgroundColor The background Color of the data on the graph
+     * @param {string} borderColor The border Color of the the data on the graph
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: missing return

File: addons/website/static/src/js/editor/snippets.options.js [hunk 21]
```diff
@@ -1940,4 +1941,450 @@ options.registry.SnippetMove = options.Class.extend({
...
+    _isStackableChart: function () {
+        return ['bar', 'horizontalBar'].includes(this.$target[0].dataset.type);
+    },
+    /**
+     * Return the number of column minus header and button
+     * @private
+     */
+    _getColumnCount: function () {
+        return this.tableEl.rows[0].cells.length - 2;
+    },
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: missing `@returns`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 22]
```diff
@@ -1940,4 +1941,450 @@ options.registry.SnippetMove = options.Class.extend({
...
+        } else {
+            this.colorPaletteSelectedInput = this.tableEl.querySelector('th input');
+        }
+    },
+    /**
+     * Return a random hexadecimal color.
+     *
+     * @private
+     */
+    _randomColor: function () {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Missing `@returns`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 23]
```diff
@@ -1940,4 +1941,450 @@ options.registry.SnippetMove = options.Class.extend({
...
+        this._addRow();
+        this._reloadGraph();
+        this._displayRemoveRowButton();
+        this.updateUI();
+    },
+    /**
+     * Remove the column and show the remove button of the next column or the last if no next.
+     *
+     * @private
+     * @param {*} ev
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Event (for all handlers)

**Comment by fja-odoo** [comment]
> Quickly change that and I'll merge ;)

Done !

**Comment by qsm-odoo** [comment]
@robodoo r+

**Pull request closed without merging**